### PR TITLE
Examine file ctime when checking if files have changed.

### LINKED ISSUE #2179: Restic uses `mtime` to detect file changes, which can miss changes.
Output of `restic version`
--------------------------
restic 0.9.4 compiled with go1.11.4 on linux/amd64

How did you run restic exactly?
-------------------------------
see below for the exact commands I used. Repo is local, and no other arguments are given to the backup command.

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Local.

Expected behavior
-----------------
Follow the script, and expect restic to backup the given file after it has been modified.

Actual behavior
---------------
"Files:   0 new,  0 changed,   4 unmodified"

Steps to reproduce the behavior
-------------------------------
```sh
echo "Hello world" > a.txt
echo "hELLO WORLD" > b.txt
touch stamp
cat a.txt > hello.txt
touch -r stamp hello.txt
restic -r /tmp/test-repo -p a.txt init
restic -r /tmp/test-repo -p a.txt backup .
sleep 10
cat b.txt > hello.txt
touch -r stamp hello.txt
restic -r /tmp/test-repo -p a.txt backup .
```

Do you have any idea what may have caused this?
-----------------------------------------------
`mtime` should not be used to determine if a file needs to be backed up, `ctime` should be used. Worst case with `ctime` is that restic will needlessly read/hash the file to determine if it has changed. By using `mtime` it can skip backing up a file.

I have seen the Debian package manager, specifically, replace a file with a different file, and put the mtime back to the same value.

If I add `-f` to the backup command, the file will indeed be backed up.

Do you have an idea how to solve the issue?
-------------------------------------------
Ideally, use `ctime`. Maybe use both, or provide an option, but `ctime` is really what should be used. It would probably have to revert to `mtime` on a filesystem that doesn't have a `ctime`.

Did restic help you or made you happy in any way?
-------------------------------------------------
It makes me happy that it otherwise seems to reliably back up my files.

### PR DESCRIPTION
**Pull request by cbane**
What is the purpose of this change? What does it change?
--------------------------------------------------------

This makes restic compare file ctimes (in addition to all of the existing checks) when it is deciding if a file has been modified.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Yes, closes #2179.

Checklist
---------

- [X] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [X] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [X] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [X] I have run `gofmt` on the code in all commits
- [X] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [X] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by cbane** [comment]
It looks like this change is breaking #2205 (or at least its unit tests). How do we want these to interact? The only idea I've had so far is disabling the ctime check if the ignore inode flag is on, but I'm open to anything that makes sense.

**Comment by d3zd3z** [comment]
The only idea I've had so far is disabling the ctime check if the ignore inode flag is on, but I'm open to anything that makes sense.

The `ctime` value really only makes sense on the exact same inode, so this seems to me like a reasonable way of handling this.

**Review by fd0 — CHANGES_REQUESTED** [review]
Thanks for taking a stab at this! This is a change which may leave users wondering what's happening, so we need to be extra careful to address this in the changelog. As far as I understood, there are some situations in which restic did not read the contents of the file, but does now (file with different content, but same length and `mtime` and inode). So for people depending on this (there are always people who depend on any behavior, no matter how obscure) will see that restic reads and hashes more files than before. Also, when metadata is changed, the `ctime` is updated and the file is re-read.

In my opinion, being correct and re-reading the contents if in doubt is the better strategy, so I'm in favor for this. Maybe we should add a paragraph to the changelog that people should open an issue if this breaks anything for them, so we can decide if we need a flag for it?

We need to clarify the following things before it can be merged:
 * Is it a bug or an enhancement (important for the changelog entry)? I'm currently leaning towards the latter, but I'm open to arguments.
 * The code needs to be rebased on master, the conflict with the ignore-inode option needs to be resolved
 * When `--ignore-inode` is turned on, I think the `ctime` should not be considered so I agree with @d3zd3z

**Comment by cbane** [comment]
OK, I rebased against master, and it now skips the ctime check when the `--ignore-inode` option is given. I also updated the text for the changelog entry.

File: changelog/unreleased/issue-2179 [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+Previously, restic only checked a file's mtime (along with other non-timestamp
+metadata) to decide if a file has changed. This could cause restic to not notice
+that a file has changed (and therefore continue to store the old version, as
+opposed to the modified version) if something edits the file and then resets the
+timestamp. Restic now also checks the ctime of files, so any modifications to a
+file should be noticed, and the modified file will be backed up. The ctime check
+will be disabled if the --ignore-inode flag was given.
+
+If this change causes problems for you, please open an issue, and we can look in
```
Comment (fd0) [inline changelog/unreleased/issue-2179]: I love the wording!

File: internal/archiver/archiver_test.go [hunk 2]
```diff
@@ -576,6 +577,19 @@ func TestFileChanged(t *testing.T) {
 				save(t, filename, defaultContent)
 			},
 		},
+		{
+			Name: "new-content-same-timestamp",
```
Comment (fd0) [inline internal/archiver/archiver_test.go]: Table-driven tests are awesome, easily extendable for contributors! :)

**Review by fd0 — APPROVED** [review]
LGTM, thanks!

**Comment by DurvalMenezes** [comment]
Please excuse me, but should'nt we be giving more warning to users about this problem? Please see my rationale [here](https://github.com/restic/restic/issues/2179#issuecomment-495654982).

**Pull request merged**